cpp/src/arrow/c/abi.h

```diff
@@ -0,0 +1,65 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ARROW_FLAG_DICTIONARY_ORDERED 1
+#define ARROW_FLAG_NULLABLE 2
+#define ARROW_FLAG_MAP_KEYS_SORTED 4
+
+struct ArrowSchema {
+  // Array type description
+  const char* format;
+  const char* name;
+  const char* metadata;
+  int64_t flags;
```

Comment:
The most compelling reason to use bit flags is that we can add new flags later if needed without breaking the ABI / changing the size of the struct, it isn't really about saving space. See e.g. the flags used in the Python Buffer protocol in https://www.python.org/dev/peps/pep-3118/

Reply:
That's flags versus independent items--I was thinking of a pointer to array of booleans. isn't constrained to a certain number of bits and avoids people doing opaque bit comparisons. Same properties as flag in terms of forward compatibility.

Reply:
I guess the argument against is knowing whether you can check for a certain flag. That kind of goes to the format version I suppose...

it's just kind of arbitrary to do 64 bits like this. why not something larger, or smaller?

And if we know some items now, why put them in flags as opposed to waiting until we need to add things later (and just having an empty flags for now).

Reply:
My other comments are more important, it just seems like a little bit of early optimization to put the known flags in here. (It may be common in python but it's viewed less positively by other communities.)

Reply:
It's not an optimization, it's really idiomatic in C to have flags as individual bits in an integer field. Using a bool array would actually be weird.

Reply:
Indeed, the idiomatic pattern is AFAIK to use the platform `int` but I presume we use `int64_t` here so that the ABI is the same across 32-bit and 64-bit systems